applications/matter_bridge/src/bridged_devices_creator.cpp

```diff
@@ -0,0 +1,384 @@
+{
+	CHIP_ERROR err;
+
+	/* Check if there is already existing provider for given address.
+	 * In case it is, the provider was either connected or recovered before. All what needs to be done is creating
+	 * new Matter endpoint. In case there isn't, the provider has to be created.
+	 */
+	BLEBridgedDeviceProvider *provider = BLEConnectivityManager::Instance().FindBLEProvider(btAddress);
+	if (provider) {
+		return AddMatterDevice(deviceType, nodeLabel, provider, index, endpointId);
+	}
+
+	/* The EnvironmentalSensor is a dummy device type, so it is not saved in the persistent storage. For now the
+	 * only BLE service handling temperature and humidity device types is environmental sensor, so it's necessary to
+	 * perform mapping. */
+	if (deviceType == MatterBridgedDevice::DeviceType::TemperatureSensor ||
+	    deviceType == MatterBridgedDevice::DeviceType::HumiditySensor) {
+		provider = static_cast<BLEBridgedDeviceProvider *>(
+			CreateBleProvider(MatterBridgedDevice::DeviceType::EnvironmentalSensor));
+	} else {
+		provider = static_cast<BLEBridgedDeviceProvider *>(CreateBleProvider(deviceType));
+	}
+
+	if (!provider) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	err = BLEConnectivityManager::Instance().AddBLEProvider(provider);
+	VerifyOrExit(err == CHIP_NO_ERROR, );
+
+	/* There are two cases for creating a new device:
+	 * - connecting new device for the first time - the index and endpoint id are not relevant, the new Bluetooth LE
+	 * connection has to be created and the operation confirmed by connected callback.
+	 * - recovering a device that was connected in the past after reboot - the index and endpoint id are read from
+	 * storage, the device has to be re-connected without confirming its existence by callback.
+	 */
+	if (index.HasValue() && endpointId.HasValue()) {
+		provider->InitializeBridgedDevice(btAddress, nullptr, nullptr);
+		BLEConnectivityManager::Instance().Recover(provider);
+		err = AddMatterDevice(deviceType, nodeLabel, provider, index, endpointId);
+	} else {
+		/* The device object can be created once the Bluetooth LE connection will be established. */
+		BluetoothConnectionContext *context = chip::Platform::New<BluetoothConnectionContext>();
+
+		VerifyOrExit(context, err = CHIP_ERROR_NO_MEMORY);
+
+		chip::Platform::UniquePtr<BluetoothConnectionContext> contextPtr(context);
+		contextPtr->deviceType = deviceType;
+
+		if (nodeLabel) {
+			strcpy(contextPtr->nodeLabel, nodeLabel);
+		}
+
+		contextPtr->provider = provider;
+		contextPtr->index = index;
+		contextPtr->endpointId = endpointId;
+		contextPtr->address = btAddress;
+
+		provider->InitializeBridgedDevice(btAddress, BluetoothDeviceConnected, contextPtr.get());
+		err = BLEConnectivityManager::Instance().Connect(provider);
+
+		if (err == CHIP_NO_ERROR) {
+			contextPtr.release();
+		}
+	}
+
+exit:
+
+	if (err != CHIP_NO_ERROR) {
+		BLEConnectivityManager::Instance().RemoveBLEProvider(btAddress);
+		delete provider;
```

Comment:
I suspect there is a chance or double deletion of `provider` when the `AddBridgedDevices` fails to add Matter device and any of the Matter device allocations fails. Therefore I suggest to choose one place/method which is responsible for deleting of the objects and return from `AddBridgedDevices` prematurely in case of allocation fails.

Reply:
Right, I removed deleting provider from the `AddMatterDevice`. Let's make the method that allocates a provider take care of deleting it (`CreateDevice`).